### ISSUE TITLE
`dead_code` lint doesn't treat uses in underscore-prefixed names as real uses

### ISSUE THREAD

**Issue by AaronKutch** [rust-lang/rust#80854]
I am making a crate with practical custom DSTs (If there is a will, there is a way). I have a number of constants which determine layouts at compile time or stop compilation in case something went wrong.

```rust
const fn hello() -> usize {
    0
}

const _ASSERT_ASSUMPTIONS: () = {
    let _ = ["Assert that the value is 0"][hello()];
};
```
However, the compiler seems to think that "hello" is unused.

**Comment by camelid** [comment]
I think this behavior is "correct". The reason the compiler is warning that `hello` is unused but not that `_ASSERT_ASSUMPTIONS` is unused is that neither of them are actually used. The only reason you don't get a warning for `_ASSERT_ASSUMPTIONS` is that it starts with an underscore.

**Comment by camelid** [comment]
Note that this behavior is not limited to constants; it also happens with:

```rust
fn hello() {}

fn _goodbye() { hello() }
```

```
warning: function is never used: `hello`
 --> src/lib.rs:1:4
  |
1 | fn hello() {}
  |    ^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: 1 warning emitted
```

**Comment by AaronKutch** [comment]
Is this just a case where a manual `#[allow(warnings)]` is what should be done?

**Comment by camelid** [comment]
I don't know what your code looks like, so I can't tell you for sure. Why is `hello` (or the equivalent function in your actual code) only used in the assertion? I would expect it to be public if it's intended to be used externally.

**Comment by AaronKutch** [comment]
It was an extremely reduced example.  It was an associated function marked `pub(crate)` on a struct. It was an important immutable constructor (the mutable constructor happened to be the only one used).

**Comment by AaronKutch** [comment]
What makes this look like a bug worth fixing to me is that I wouldn't expect compilation to ever fail if I removed a function warned by the dead_code lint.

**Comment by skysch** [comment]
Dead code is code that is never called, not code that is never referenced.

```
fn b() { a() }
fn a() { b() }
```

This will produce dead_code warnings for both functions, and removing either results in a compilation failure. (And that is probably what it should do, as these functions clearly do nothing by themselves.) It seems like you're actually expecting the `_` prefix to apply transitively.

**Comment by AaronKutch** [comment]
That reasoning makes sense to me. I suppose my use case is pretty rare, even though I think I have encountered it more than once. If no one else thinks it is a problem, I can close this issue.
